video/out/vo_gpu_next.c

```diff
@@ -1519,51 +1521,56 @@ static void load_cache_files(struct priv *p)
         }
         talloc_free(shader_cache);
     }
-    if (icc_cache) {
-        if (!same_cache) {
-            FILE *cache = fopen(icc_cache, "rb");
-            if (cache) {
-                int ret = pl_cache_load_file(p->icc_cache, cache);
-                fclose(cache);
-                if (ret < 0)
-                    MP_WARN(p, "Failed loading cache from %s\n", icc_cache);
-            }
+    if (icc_cache && !same_cache) {
+        FILE *cache = fopen(icc_cache, "rb");
+        if (cache) {
+            int ret = pl_cache_load_file(p->icc_cache, cache);
+            fclose(cache);
+            if (ret < 0)
+                MP_WARN(p, "Failed loading cache from %s\n", icc_cache);
         }
-        talloc_free(icc_cache);
     }
+    talloc_free(icc_cache);
 }
 
 static void save_cache_files(struct priv *p)
 {
+    void *ta_ctx = talloc_new(NULL);
     char *icc_cache = get_cache_file(p, "icc");
     char *shader_cache = get_cache_file(p, "shader");
+    talloc_steal(ta_ctx, icc_cache);
+    talloc_steal(ta_ctx, shader_cache);
+
     bool same_cache = false;
     if (icc_cache && shader_cache)
         same_cache = strcmp(icc_cache, shader_cache) == 0;
-    if (shader_cache) {
-        FILE *cache = fopen(shader_cache, "wb");
-        if (cache) {
-            int ret = pl_cache_save_file(p->shader_cache, cache);
-            if (same_cache)
-                pl_cache_save_file(p->icc_cache, cache);
-            fclose(cache);
-            if (ret < 0)
-                MP_WARN(p, "Failed saving cache to %s\n", shader_cache);
-        }
-        talloc_free(shader_cache);
-    }
-    if (icc_cache) {
-        if (!same_cache) {
-            FILE *cache = fopen(icc_cache, "wb");
-            if (cache) {
-                int ret = pl_cache_save_file(p->icc_cache, cache);
-                fclose(cache);
-                if (ret < 0)
-                    MP_WARN(p, "Failed saving cache to %s\n", icc_cache);
-            }
+    for (int i = 0; i < 2; i++) {
+        const char *target_file = i == 0 ? shader_cache : icc_cache;
+        pl_cache target_cache = i == 0 ? p->shader_cache : p->icc_cache;
+
+        if (!target_file)
+            continue;
+        char *tmp = talloc_asprintf(ta_ctx, "%s~", target_file);
+        FILE *cache = fopen(tmp, "wb");
+        if (!cache)
+            continue;
+        int ret = pl_cache_save_file(target_cache, cache);
+        if (same_cache)
+            ret += pl_cache_save_file(p->icc_cache, cache);
+        fclose(cache);
+        if (ret == 0)
+            unlink(tmp);
```

Comment:
I would merge the unlink with error case, it read weird that rename is interlocked between two unlinks.

Reply:
I can't just move it, after the rename `ret` is zero which would cause a wrong `unlink` call.
I don't think there's a more compact way to write the logic that needs to be here.

Reply:
Before that. `ret == pl_cache_save_file(target_cache, cache)` if valid cache save when there are no object saved and should be respected. Save empty (with only heder) cache file.

Reply:
You want to keep the empty cache files?

Reply:
Yes, why not? If there are no object to save, meaning that libplacebo decided to invalidate everything, why keep stale file?

Reply:
IIRC it will still write the empty header, so it should not return 0 ever.

Reply:
It return number of object written, not bytes.

Reply:
Also, does this `same_file` logic work? As far as I can see it will just write 2nd header and objects, but cache load will load only first one, no?

Reply:
What I could do is also `unlink(target_file)` in that case. Would unfortunately make the entire operation a bit more racy.

Reply:
I don't think it is necessary to unlink target file. What's wrong with letting libplacebo save the file as if it decides to save only header with 0 objects? Maybe it should be fixed on libplacebo side if this behavior is not desirable?

Reply:
Well whatever, I'll change it.